src/include/simeng/OS/SimOS.hh

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include "simeng/Config.hh"
+#include "simeng/Core.hh"
+#include "simeng/OS/SyscallHandler.hh"
+#include "simeng/SpecialFileDirGen.hh"
+#include "simeng/arch/Architecture.hh"
+#include "simeng/arch/aarch64/Architecture.hh"
+#include "simeng/arch/aarch64/Instruction.hh"
+#include "simeng/arch/riscv/Architecture.hh"
+#include "simeng/arch/riscv/Instruction.hh"
+#include "simeng/memory/SimpleMem.hh"
+#include "simeng/span.hh"
+
+// Forward declare RegressionTest class so that it can be declared a friend.
+class RegressionTest;
+
+namespace simeng {
+namespace OS {
+
+// Program used when no executable is provided; counts down from
+// 1024*1024, with an independent `orr` at the start of each branch.
+static uint32_t hex_[8] = {
+    0x320C03E0,  // orr w0, wzr, #1048576
+    0x320003E1,  // orr w0, wzr, #1
+    0x71000400,  // subs w0, w0, #1
+    0x54FFFFC1,  // b.ne -8
+                 // .exit:
+    0xD2800000,  // mov x0, #0
+    0xD2800BC8,  // mov x8, #94
+    0xD4000001,  // svc #0
+};
+
+/** A simple, lightweight Operating System kernel based on Linux to emulate
+ * syscalls and manage process execution. */
+class SimOS {
+ public:
+  /** Construct a SimOS object. */
+  SimOS(std::string executablePath, std::vector<std::string> executableArgs,
+        std::shared_ptr<simeng::memory::Mem> mem);
+
+  /** Tick SimOS. */
+  void tick();
+
+  /** Get a process with specified TID. */
+  std::shared_ptr<Process> getProcess(uint64_t TID) const;
+
+  /** Get shared_ptr to syscallHandler instance. */
+  std::shared_ptr<SyscallHandler> getSyscallHandler() const {
+    return syscallHandler_;
+  }
+
+  /** Register a core with the OS to enable process scheduling. */
+  void registerCore(std::shared_ptr<simeng::Core> core) {
+    cores_.push_back(core);
+  }
+
+  /** Check if OS has halted. */
+  bool hasHalted() const { return halted_; };
+
+  /** Set up friend class with RegressionTest to enable exclusive access to
+   * private functions. */
+  friend class ::RegressionTest;
+
+ private:
+  /** Create the initial SimOS Process running above this kernel from command
+   * line arguments.
+   * Empty command line arguments denote the usage of hardcoded instructions
+   * held in the hex_ array.*/
+  void createInitialProcess();
+
+  /** Construct the special file directory. */
+  void createSpecialFileDirectory() const;
+
+  /** The path of user defined Executable. */
+  std::string executablePath_ = DEFAULT_STR;
+
+  /** The runtime arguments of the user defined executable. */
+  std::vector<std::string> executableArgs_ = {};
+
+  /** The list of user-space processes running above the OS kernel. */
+  // std::vector<std::shared_ptr<Process>> processes_ = {};
+
+  std::unordered_map<uint64_t, std::shared_ptr<Process>> processes_ = {};
```

Comment:
The only alternative to using a smart pointer would be to use c-style pointers (i.e. `Process*`). This is because a process most likely exists in 2 data structures withing SimOS at a given time (`processes_` and `waitingProcs_` XOR `scheduledProcs_`), and you are not able to store references in a queue

Reply:
If we can't store refs in the queue, this is a valid use of smart pointers. Also from a performance perspective, the number of times we move processes around compare to ticking the core(s) is minute.

Reply:
Agreed, a const ref would be better.